### PR TITLE
7535 extract partition service

### LINKED ISSUE #7535: Extract PartitionService from Atomix
**Description**

Goal of this issue is to extract the `PartitionService` from the `Atomix` class.

The steps have already been attempted in https://github.com/camunda-cloud/zeebe/pull/7381 (except for refactoring tests)

Out of scope for this issue is to change the way partitions are created and registered. This will be changed in a later issue.

**Implementation Steps** 
* Crreate new class `PartitionManager`
* Move logic that deals with `PartitionService` from `Atomix` to `PartitionManager`
* Adjust `AtomixFacotry, AtomixBuilder. AtomixConfig, AtomixRule`
* Remove `PartitionService` from `ClusterServcies`
* Create `PartitionManager` in `Broker#partitionsStep`
* Create `SnapshotStoreFactory` in `Broker#partitionStep`
* Refactor test classes

### PR DESCRIPTION
**Pull request by pihme**
## Description

This PR extracts the `PartitionService` from the `Atomix` class and encapsulates it in a new class `PartitionManager`.

As a consequence of this change, the `Atomix` class, related classes and the whole `atomix-core` module are removed. Basically, without the `PartitionService` there was no longer a strong reason for these classes to exist.

Note that this is only the first step in the development of `PartitionManager`. Later on it will offer more interfaces to the outside. Probably at that time it will also become an `Actor`. But for now it is just a simple class to wrap the `PartitionService`

### Review Hints 
Mainly commit by commit. Most commits are narrow in scope. Except for the one central big commit, where the actual extraction happens. 

## Related issues

closes #7535 

## Definition of Done

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [X] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.agrona.IoUtil;
+
+public class PartitionManager {
+  protected volatile CompletableFuture<Void> closeFuture;
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java:24]: :question: Are we expecting to sub-class this?

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+              maxMessageSize, segmentSize));
+    }
+
+    partitionGroupBuilder.withSegmentSize(segmentSize);
+
+    return partitionGroupBuilder.build();
+  }
+
+  public synchronized CompletableFuture<Void> start() {
+    if (closeFuture != null) {
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java:43]: :wrench: (If we do make this an actor, then you can ignore this suggestion) I assume we're using synchronized here to ensure stop/start are mutually exclusive. I would propose to use a specific lock instead of the object's, as anybody adding more `synchronized` methods/blocks in the future might be blocking these calls unintentionally.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      final ReceivableSnapshotStoreFactory snapshotStoreFactory,
+      final ClusterMembershipService membershipService,
+      final ClusterCommunicationService communicationService) {
+
+    partitionGroup = buildRaftPartitionGroup(configuration, snapshotStoreFactory);
+
+    partitions = buildPartitionService(membershipService, communicationService);
+  }
+
+  // TODO remove static modifier
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java]: :question: What's the plan for this later? Will it stay in the partition manager? Will we have a partition group builder (well not builder, that's an overloaded word, but you get what I mean)?

File: atomix/cluster/src/test/java/io/atomix/cluster/AtomixClusterRule.java [hunk 4]
```diff
@@ -88,7 +88,7 @@ public AtomixBuilder buildAtomix(
                 })
             .collect(Collectors.toList());
 
-    return Atomix.builder(new AtomixConfig())
+    return new AtomixClusterBuilder(new ClusterConfig())
```
Comment (npepinpe) [inline atomix/cluster/src/test/java/io/atomix/cluster/AtomixClusterRule.java:91]:
```suggestion
    return AtomixCluster.builder()
```

File: broker/src/main/java/io/camunda/zeebe/broker/clustering/ClusterServices.java [hunk 5]
```diff
@@ -7,47 +7,42 @@
...
 import io.atomix.cluster.ClusterMembershipService;
 import io.atomix.cluster.messaging.ClusterCommunicationService;
 import io.atomix.cluster.messaging.ClusterEventService;
 import io.atomix.cluster.messaging.MessagingService;
-import io.atomix.core.Atomix;
-import io.atomix.primitive.partition.ManagedPartitionGroup;
 import java.util.concurrent.CompletableFuture;
 
-public final class ClusterServices {
+public class ClusterServices {
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/clustering/ClusterServices.java:17]: :thought_balloon: I can imagine making this an interface later, and having the `AtomixCluster` based adapter for it. I would do it now personally, as it's little effort and I think it has some obvious benefits for testing, but I'm happy to leave that for later.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManagerFactory.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        clusterServices.getCommunicationService());
+  }
+
+  private static RaftPartitionGroup buildRaftPartitionGroup(
+      final BrokerCfg configuration, final ReceivableSnapshotStoreFactory snapshotStoreFactory) {
+
+    final DataCfg dataConfiguration = configuration.getData();
+    final String rootDirectory = dataConfiguration.getDirectory();
+    try {
+      FileUtil.ensureDirectoryExists(new File(rootDirectory).toPath());
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManagerFactory.java:50]: :wrench: I know you didn't add it, but we can already simplify this to:

```java
FileUtil.ensureDirectoryExists(Paths.get(rootDirectory));
```

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManagerFactory.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    } catch (final IOException e) {
+      throw new UncheckedIOException("Failed to create data directory", e);
+    }
+
+    final File raftDirectory = new File(rootDirectory, GROUP_NAME);
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManagerFactory.java:55]: :wrench: Similarly, we can reuse the path constructed above and just do `rootPath.resolve(GROUP_NAME)` to get the new path. I would also propose we replace the `IoUtil.ensureDirectoryExists` usage with `FileUtil.ensureDirectoryExists` as above - we had an issue to replace all their usages, but I guess we missed one. The `IoUtil` variant unfortunately hides the error - it simply says if it failed or not, but you don't know why, which is why we decided recently to always go for the `FileUtil` variant.

I know it's a bit out of scope since you just extracted this code, so I'm happy to have this in a follow up.

File: broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java [hunk 8]
```diff
@@ -100,19 +101,21 @@
   private volatile boolean brokerStarted = false;
   private final HealthMonitor healthMonitor;
   private final MemberId nodeId;
-  private final PartitionGroup partitionGroup;
 
-  public BrokerHealthCheckService(
-      final BrokerInfo localBroker, final PartitionGroup partitionGroup) {
-    this.partitionGroup = partitionGroup;
+  public BrokerHealthCheckService(final BrokerInfo localBroker) {
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java:105]: :question: IIRC, the service knew that it was not ready because it was waiting on some partition information. So now to use it, you first have to register the partition manager at which point its status can then be changed? Is it statically not ready until then?

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java [hunk 9]
```diff
@@ -22,8 +22,8 @@
 public class PartitionManager {
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java:22]: :question: Is there a reason not to make this final? If it's for mocking, we can already introduce an interface, as tbh, I expect this will become an interface anyway (though we may not yet know the final API).

File: broker/pom.xml [hunk 10]
```diff
@@ -1,261 +1,250 @@
...
     </dependency>
-
     <dependency>
       <groupId>io.camunda</groupId>
-      <artifactId>zeebe-protocol-impl</artifactId>
+      <artifactId>zeebe-atomix-cluster</artifactId>
+      <version>${project.version}</version>
+      <scope>test</scope>
+      <classifier>tests</classifier>
+      <type>test-jar</type>
```
Comment (npepinpe) [inline broker/pom.xml:56]: :question: What do we need from the test JAR? In general I would avoid using test JAR as that breaks module boundaries, but I understand sometimes it's temporarily necessary.

**Review by npepinpe — COMMENTED** [review]
Thanks! It was a bit difficult to review for me because it seems like it's laying the foundation for further issues, and it made me realize I'm still not 100% sure of where you want to go, so I had more questions than usual I guess :smile: 

Some random notes that didn't fit in a specific line:

- :thought_balloon: We should address the package structure, but probably more towards the end. I'm not sure about the `partitioning` package, not with the structure I'm thinking of, but I also realize we haven't really written it down. I'm adding this here so I don't forget, but also so you're aware and in the chance you want to discuss it or have also been thinking about it.
- We should avoid as much as possible using static objects, as shared state makes it harder to reason about and debug concurrent issues, and will impede parallelization of tests.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                    return null;
+                  });
+    }
+
+    return closeFuture;
+  }
+
+  @Override
+  public String toString() {
+    return toStringHelper(this).add("partitionGroup", partitionGroup).toString();
```
Comment (npepinpe) [inline broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java:72]: :upside_down_face: I know this is copied code, but using `toStringHelper` is an Atomix thing, and in Zeebe we typically use just the normal string concatenation of IntelliJ by convention. I realize this isn't in the style guide, so I'll update that.

**Review by npepinpe — COMMENTED** [review]
A few more notes: our style guide specifically mentions that it's not applied in Atomix since it was a hard fork, but as we migrate this code and take ownership, we should update the style to reflect Zeebe's so we remain consistent.

File: broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java (hunk 8 above)
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java:105]: Didn't get this comment. What is your concern here?

File: broker/pom.xml (hunk 10 above)
Comment (pihme) [reply broker/pom.xml:56]: The rule and the Noop snapshot classes

**Comment by pihme** [comment]
> thought_balloon We should address the package structure, but probably more towards the end. I'm not sure about the partitioning package, not with the structure I'm thinking of, but I also realize we haven't really written it down. I'm adding this here so I don't forget, but also so you're aware and in the chance you want to discuss it or have also been thinking about it.

I have not been thinking about it in detail, no. The only thing is that if we want to use `PartitionManager` in broker and gateway, then it either needs to move into a Maven project that both depend upon already. Which sort of suggests to make it a Maven project of its own. This also fits into something you suggested when we discussed "taking over ownership of atomix". At that time you suggested that each of the services could potentially become its own module. 

So my general thinking so far is:
* keep it as a package for now
* Move it to own Maven module when we do the part about gateway alongside broker, but on the same level of abstraction
* Maybe, in the future (but definitely out of the scope of this refactoring) move more Raft stuff out of Atomix cluster and into the new partitions module.

**Comment by pihme** [comment]
> We should avoid as much as possible using static objects, as shared state makes it harder to reason about and debug concurrent issues, and will impede parallelization of tests.

Agreed. Not sure which bit caught your eye, but `PartitionManagerFactory` comes immediately to mind. Some of it was to make reviewing easier. Part of it was to make the whole change set as small as possible. You can rest assured that I will make it unstatic, as soon as it bothers me at some place :-)

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java (hunk 2 above)
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java:43]: It's a good point. I will ignore it for now, though. I want to wait and see how this class evolves, and then revisit synchronization again.

File: broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java (hunk 3 above)
Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/partitioning/PartitionManager.java]: Outdated

File: atomix/cluster/src/test/java/io/atomix/cluster/AtomixClusterRule.java (hunk 4 above)
Comment (pihme) [reply atomix/cluster/src/test/java/io/atomix/cluster/AtomixClusterRule.java:91]: Class no longer exists

**Comment by pihme** [comment]
> I'm still not 100% sure of where you want to go

Neither am I :smile:

File: broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java (hunk 8 above)
Comment (npepinpe) [reply broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java:105]: Not a concern, just a question :sweat_smile: I'm trying to understand the semantics of the class now. So let me rephrase - what's the health status of the service _before_ I register the partition manager?

Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java:105]: Good question. Let me write a test for it

Comment (pihme) [reply broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/BrokerHealthCheckService.java:105]: the health status, ready status and started status are all false before the partition manager is registered.

**Review by npepinpe — APPROVED** [review]
:rocket: 

Thanks for answering all my questions :slightly_smiling_face: Looks good to me, any changes I would ask I think will be addressed soon. I would like to take some time at the end to maybe see if we can't get rid of the test JAR dependency. Seems like a code-smell (or whatever the architectural equivalent for dependencies/module boundaries)

**Comment by pihme** [comment]
>  I would like to take some time at the end to maybe see if we can't get rid of the test JAR dependency

Could you create an issue for that?
